modules/libarchive/3.7.3/patches/build_with_bazel.patch

```diff
@@ -0,0 +1,1805 @@
++#define HAVE_SYS_TIME_H 1
++
++#define HAVE_CHROOT 1
++#define HAVE_CTIME_R 1
++#define HAVE_DIRFD 1
++#define HAVE_FCHDIR 1
++#define HAVE_FCHMOD 1
++#define HAVE_FCHOWN 1
++#define HAVE_FCNTL 1
++#define HAVE_FDOPENDIR 1
++#define HAVE_FGETXATTR 1
++#define HAVE_FLISTXATTR 1
++#define HAVE_FORK 1
++#define HAVE_FSETXATTR 1
++#define HAVE_FSTATAT 1
++#define HAVE_FSTATFS 1
++#define HAVE_FSTATVFS 1
++#define HAVE_FUTIMENS 1
++#define HAVE_FUTIMES 1
++#define HAVE_GETEUID 1
++#define HAVE_GETGRGID_R 1
++#define HAVE_GETGRNAM_R 1
++#define HAVE_GETPWNAM_R 1
++#define HAVE_GETPWUID_R 1
++#define HAVE_GETXATTR 1
++#define HAVE_GMTIME_R 1
++#define HAVE_GRP_H 1
++#define HAVE_LANGINFO_H 1
++#define HAVE_LCHOWN 1
++#define HAVE_LGETXATTR 1
++#define HAVE_LINK 1
++#define HAVE_LISTXATTR 1
++#define HAVE_LLISTXATTR 1
++#define HAVE_LOCALTIME_R 1
++#define HAVE_LSETXATTR 1
++#define HAVE_LSTAT 1
++#define HAVE_LUTIMES 1
++#define HAVE_MKFIFO 1
++#define HAVE_MKNOD 1
++#define HAVE_MKSTEMP 1
++#define HAVE_NL_LANGINFO 1
++#define HAVE_OPENAT 1
++#define HAVE_PIPE 1
++#define HAVE_POLL 1
++#define HAVE_POLL_H 1
++#define HAVE_POSIX_SPAWNP 1
++#define HAVE_PWD_H 1
++#define HAVE_READLINK 1
++#define HAVE_READLINKAT 1
++#define HAVE_REGEX_H 1
++#define HAVE_SELECT 1
++#define HAVE_SIGACTION 1
++#define HAVE_SPAWN_H 1
++#define HAVE_STATFS 1
++#define HAVE_STAVTFS 1
++#define HAVE_STRERROR_R 1
++#define HAVE_STRUCT_STAT_ST_BLKSIZE 1
++#define HAVE_STRUCT_TM_TM_GMTOFF 1
++#define HAVE_SYMLINK 1
++#define HAVE_SYS_IOCTL_H 1
++#define HAVE_SYS_MOUNT_H 1
++#define HAVE_SYS_POLL_H 1
++#define HAVE_SYS_SELECT_H 1
++#define HAVE_SYS_UTSNAME_H 1
++#define HAVE_SYS_VFS_H 1
++#define HAVE_SYS_WAIT_H 1
++#define HAVE_TIMEGM 1
++#define HAVE_UNISTD_H 1
++#define HAVE_UNSETENV 1
++#define HAVE_UTIMENSAT 1
++#define HAVE_UTIMES 1
++#define HAVE_VFORK 1
++#define LSTAT_FOLLOWS_SLASHED_SYMLINK 1
++#define STRERROR_R_CHAR_P 1
++
++#define HAVE_ZLIB_H 1
++#define HAVE_ZSTD_H 1
++#define HAVE_LIBZSTD 1
++#define HAVE_LIBZSTD_COMPRESSOR 1
++#define HAVE_LIBLZMA 1
++#define HAVE_LZMA_H 1
++#define HAVE_LZMA_STREAM_ENCODER_MT 1
++#define HAVE_LIBBZ2 1
++#define HAVE_BZLIB_H 1
++#define HAVE_LIBLZ4 1
++#define HAVE_LZ4_H 1
++#define HAVE_LZ4HC_H 1
++
++#ifdef __APPLE__
++#define MAC_OS_X_VERSION_MIN_REQUIRED 1060
++#define MAC_OS_X_VERSION_MAX_ALLOWED 1070
++#define HAVE_STRUCT_STAT_ST_BIRTHTIME 1
++#define HAVE_SYS_CDEFS_H 1
++#endif
++
++#ifdef __linux__
++#if (defined(__GNUC__) && !defined(__clang__))
++#define HAVE_FUTIMESAT 1
++#endif
++#define HAVE_LINUX_FS_H 1
++#define HAVE_LINUX_MAGIC_H 1
++#define HAVE_LINUX_TYPES_H 1
++#define HAVE_PATHS_H 1
++#define HAVE_STRUCT_STAT_ST_MTIM_TV_NSEC 1
++#define HAVE_SYS_XATTR_H 1
++#endif
+diff --git a/libarchive_bazel_windows_config.h b/libarchive_bazel_windows_config.h
+new file mode 100644
+index 00000000..baf68cff
+--- /dev/null
++++ b/libarchive_bazel_windows_config.h
+@@ -0,0 +1,110 @@
++#define __LIBARCHIVE_CONFIG_H_INCLUDED 1
```

Comment:
why make this one separate vs having a single `config.h` with some ifdefs for windows? Then you wouldn't need to vary the config filename either?

not a blocker, but it would be nice to resolve this so we don't need additional changes to the config.h includes

Reply:
I just generated it with cmake and figured it had the most coverage. I could do what you're saying and add a bunch of preprocessor ifs, but it seemed very different than the 'generic' one. I could go either way 🤷

Reply:
another option is to have a `copy_file` rule that creates a `config.h` based on the right platform. I'm leaning toward doing either that or handwriting the config.h, but both of those seem preferable to patching the libarchive source to support a dynamic config filename (especially if/when https://github.com/bazelbuild/bazel-central-registry/issues/1566 lands and we can do overlays for BUILD files instead of having to do patches)

Reply:
I think doing `copy_file` is a little overkill since these are _new_ files anyways. From reading #1566 it seems that having new non-bazel files would be supported.

Reply:
Sorry, I think I could have been clearer. I agree that new files are supported, I'm just trying to avoid the need to patch the header name into the libarchive sources, which was why I suggested using the upstream name (`config.h`). `copy_file` is just one way to accomplish that, another one would be a `write_file` or `genrule` with a platform-dependent content

(BTW this comment is not blocking, just some thoughts to make future maintenance easier)

Reply:
oh we don't have to patch the header name since libarchive supports a custom path with `PLATFORM_CONFIG_H` define!

Reply:
oh I see, were you previously patching it? Not sure why I thought you were. I think it's probably still slightly less code overall to not have the platform-dependent names, but not a huge deal either way

Reply:
I did patch one spot since the test util file didn't use the `PLATFORM_CONFIG_H`, but I intend to upstream that change soon.

> I think it's probably still slightly less code overall to not have the platform-dependent names

I do agree having the 1 bazel config header would be nicer. I might give it a try, but the generated cmake one might be more accurate/maintainable.

Reply:
Yeah I think using the cmake header is probably fine, I was thinking something like this
```
copy_file(
  name = "gen_config",
  outs = "config.h",
  src = select({
    "@platforms//os:windows": "cmake_generated_windows_config.h",
    "//conditions:default": "cmake_generated_windows_config.h",
     }),
 )
 ```

Reply:
ahh I see what you're saying. Just to consolidate the files into one. I'm fine to add that change. Is there a specific reason you like having it as `config.h` instead of having the platform and 'bazel' in the name? 

Having the platform name almost makes it clearer to me which one I'm using, so I wasn't bothered by it.

I would like to avoid using `HAVE_CONFIG_H` since `libarchive` uses the config file in their public headers to avoid issues like this: https://github.com/bazelbuild/bazel-central-registry/pull/1917#issuecomment-2093984116, but I can still use `PLATFORM_CONFI_H` with it being `config.h` so that shouldn't be an issue.

Reply:
Ah I see. I was going to suggest local_defines but doesn't sound like that works here. Maybe upstream would take a patch renaming it to HAVE_LIBARCHIVE_CONFIG_H? I'm surprised non-bazel builds don't suffer from this

Anyway, what you wrote in the last comment makes sense